### PR TITLE
Throw ParseException for deserialize.

### PR DESCRIPTION
**Pull request by songy23**
Consider using ParseException instead of IOException for deserialize().

### CODE REVIEW

File: core_impl/src/main/java/com/google/instrumentation/stats/StatsSerializer.java [hunk 1]
```diff
@@ -88,18 +89,18 @@ static void serialize(StatsContextImpl context, OutputStream output) throws IOEx
 
   // Deserializes input to StatsContext based on the binary format standard.
   // The encoded tags are of the form: <version_id><encoded_tags>
-  static StatsContextImpl deserialize(InputStream input) throws IOException {
+  static StatsContextImpl deserialize(InputStream input) throws IOException, ParseException {
```
Comment (bogdandrutu) [inline core_impl/src/main/java/com/google/instrumentation/stats/StatsSerializer.java]: We discussed with the gRPC team and they suggested to use byte[] instead of Input[Output]Stream because that is what they need.

Comment (songy23) [inline core_impl/src/main/java/com/google/instrumentation/stats/StatsSerializer.java]: A problem with this: If they want to serialize() and deserialize() with byte[], for serialization they need to pass in an array that's large enough, but then there will probably be trailing 0s, and trying to deserialize that array will throw an exception. How are we supposed to resolve that?

Comment (bogdandrutu) [inline core_impl/src/main/java/com/google/instrumentation/stats/StatsSerializer.java]: I was thinking to have something like this:
byte[] serialize(StatsContext ctxt) {}
StatsContext deserialize(byte[]) {}

Comment (songy23) [inline core_impl/src/main/java/com/google/instrumentation/stats/StatsSerializer.java]: Create a separate issue for this: https://github.com/google/instrumentation-java/issues/213
Will make the code changes in another PR.

**Comment by adriancole** [comment]
Personally i dont like using exception types that are close but no cigar.
For example, parse exception was made for a different but similar purpose.
Style choice of course but i would throw something different if you cannot
parse like illegalargument

**Comment by adriancole** [comment]
Specifically this is a rather obscure text exception. It is valient to try
to reuse types but this one is a miss imho.
